File: app/src/main/java/edu/ucla/cs/er/depressionstudy/SurveyFragment.java
//package edu.ucla.cs.er.depressionstudy;
//
//
//import android.app.Activity;
//import android.content.Intent;
//import android.database.Cursor;
//import android.os.Bundle;
//import androidx.fragment.app.Fragment;
//import android.util.Log;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//import android.widget.Button;
//import android.widget.ImageView;
//import android.widget.TextView;
//
//import com.aware.Aware;
//import com.aware.Aware_Preferences;
//import com.aware.ESM;
//import com.aware.providers.ESM_Provider;
//import com.aware.ui.ESM_Queue;
//import com.aware.ui.esms.ESMFactory;
//import com.aware.ui.esms.ESM_Checkbox;
//import com.aware.ui.esms.ESM_Number;
//import com.aware.ui.esms.ESM_PAM;
//import com.aware.ui.esms.ESM_QuickAnswer;
//import com.aware.ui.esms.ESM_Radio;
//import com.quickbirdstudios.surveykit.steps.InstructionStep;
//import com.quickbirdstudios.surveykit.survey.SurveyView;
//
//
//import org.json.JSONArray;
//import org.json.JSONException;
//
//import java.util.Calendar;
//
//
///**
// * A simple {@link Fragment} subclass.
// */
//public class SurveyFragment extends Fragment {
//
//    private Button startButtonSurvey;
//    private Activity context;
//    private String esmString;
//
//    public SurveyFragment() {
//        // Required empty public constructor
//    }
//
//
//    @Override
//    public void onResume() {
//        super.onResume();
//
//        //updateCompletedTextVisibility(this.getView());
//    }
//
//    @Override
//    public View onCreateView(LayoutInflater inflater, ViewGroup container,
//                             Bundle savedInstanceState) {
//        // Inflate the layout for this fragment
//        View rootView = inflater.inflate(R.layout.fragment_survey, container, false);
//        context = getActivity();
//
//        SurveyView surveyView = rootView.findViewById(R.id.survey_view);
//
//
////        startButtonSurvey = (Button) rootView.findViewById(R.id.bt_start_survey);
////        startButtonSurvey.setOnClickListener(new Button.OnClickListener() {
////            public void onClick(View v) {
////                try {
////                    ESMFactory factory = new ESMFactory();
////
////                    ESM_QuickAnswer thank_you = new ESM_QuickAnswer();
////                    thank_you.addQuickAnswer("Close")
////                            .setInstructions("Thank you for filling the survey!")
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer sbq4 = new ESM_QuickAnswer();
////                    sbq4.addQuickAnswer("0-4 times")
////                            .addQuickAnswer("≥5 times")
////                            .setInstructions("In the last 3 months, how many times did you have insertive anal sex (you were the top) with a man who is HIV-positive when you did not use a condom?")
////                            .addFlow("0-4 times", thank_you.build())
////                            .addFlow("≥5 times", thank_you.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer sbq3 = new ESM_QuickAnswer();
////                    sbq3.addQuickAnswer(">1 time")
////                            .addQuickAnswer("1 time")
////                            .addQuickAnswer("0 times")
////                            .addQuickAnswer("Don't know")
////                            .setInstructions("In the last 3 months, how many of your male sex partners were HIV-positive")
////                            .addFlow(">1 time", sbq4.build())
////                            .addFlow("1 time", sbq4.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer sbq2 = new ESM_QuickAnswer();
////                    sbq2.addQuickAnswer("≥1 time")
////                            .addQuickAnswer("0 times")
////                            .setInstructions("In the last 3 months, how many times did you have receptive anal sex (you were the bottom) with a man when he did not use a condom?")
////                            .addFlow("≥1 time", sbq3.build())
////                            .addFlow("0 time", sbq3.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer sbq1 = new ESM_QuickAnswer();
////                    sbq1.addQuickAnswer(">10 male partners")
////                            .addQuickAnswer("6-10 male partners")
////                            .addQuickAnswer("1-5 male partners")
////                            .addQuickAnswer("0 male partners")
////                            .setInstructions("In the last 3 months, how many men have you had sex with?")
////                            .addFlow(">10 male partners", sbq2.build())
////                            .addFlow("6-10 male partners", sbq2.build())
////                            .addFlow("1-5 male partners", sbq2.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer pwid6 = new ESM_QuickAnswer();
////                    pwid6.addQuickAnswer("Yes")
////                            .addQuickAnswer("No")
////                            .setInstructions("In the past 3 months, did you visit a shooting gallery?")
////                            .addFlow("Yes", sbq1.build())
////                            .addFlow("No", sbq1.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer pwid5 = new ESM_QuickAnswer();
////                    pwid5.addQuickAnswer("Yes")
////                            .addQuickAnswer("No")
////                            .setInstructions("In the past 3 months, did you share a cooker?")
////                            .addFlow("Yes", pwid6.build())
////                            .addFlow("No", pwid6.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer pwid4 = new ESM_QuickAnswer();
////                    pwid4.addQuickAnswer("Yes")
////                            .addQuickAnswer("No")
////                            .setInstructions("In the past 3 months, did you use needles?")
////                            .addFlow("Yes", pwid5.build())
////                            .addFlow("No", pwid5.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer pwid3 = new ESM_QuickAnswer();
////                    pwid3.addQuickAnswer("Yes")
////                            .addQuickAnswer("No")
////                            .setInstructions("In the past 3 months, did you inject heroin?")
////                            .addFlow("Yes", pwid4.build())
////                            .addFlow("No", pwid4.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer pwid2 = new ESM_QuickAnswer();
////                    pwid2.addQuickAnswer("Yes")
////                            .addQuickAnswer("No")
////                            .setInstructions("In the past 3 months, did you inject cocaine?")
////                            .addFlow("Yes", pwid3.build())
////                            .addFlow("No", pwid3.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer pwid1 = new ESM_QuickAnswer();
////                    pwid1.addQuickAnswer("Yes")
////                            .addQuickAnswer("No")
////                            .setInstructions("In the past 3 months, were you in a methadone maintenance program?")
////                            .addFlow("Yes", pwid2.build())
////                            .addFlow("No", pwid2.build())
////                            .setTitle("Survey");
////
////                    ESM_QuickAnswer su1 = new ESM_QuickAnswer();
////                    su1.addQuickAnswer("Yes, methamphetamines")
////                            .addQuickAnswer("Yes, injectable drugs not prescribed")
////                            .addQuickAnswer("Yes, used both")
////                            .addQuickAnswer("Neither")
////                            .setInstructions("In the past 3 months, have you used substances such as crystal meth or injectable drugs not prescribed to you by a physician?")
////                            .addFlow("Yes, methamphetamines", sbq1.build())
////                            .addFlow("Neither", sbq1.build())
////                            .addFlow("Yes, injectable drugs not prescribed", pwid1.build())
////                            .addFlow("Yes, used both", pwid1.build())
////                            .setTitle("Survey");
////
////                    ESM_Number ageq = new ESM_Number();
////                    ageq.setTitle("Survey")
////                            .setInstructions("How old are you today?")
////                            .setSubmitButton("Next")
////                            .setExpirationThreshold(3600)
////                            .setNotificationTimeout(0)
////                            .setReplaceQueue(true);
////
////                    factory.addESM(ageq);
////                    factory.addESM(su1);
////                    //factory.addESM(sbq1);
////
////                    //Queue them
////                    //ESM.queueESM(context, factory.build());
////                    /*Intent esm = new Intent(ESM.ACTION_AWARE_QUEUE_ESM);
////                    esm.putExtra(ESM.EXTRA_ESM, factory.build());
////                    context.sendBroadcast(esm);*/
////                    ESM.queueESM(getContext(), factory.build());
////
////                    /*Intent intent_ESM = new Intent(context, ESM_Queue.class);
////                    intent_ESM.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
////                    context.startActivity(intent_ESM);*/
////                } catch (JSONException e) {
////                    e.printStackTrace();
////                }
////            }
////        });
//
//        //updateCompletedTextVisibility(rootView);
//
//        return rootView;
//    }
//
//    private boolean hasFilledSurveyToday() {
//        Aware.setSetting(context, Aware_Preferences.STATUS_ESM, true);
//        Cursor survey_data = context.getContentResolver().query(ESM_Provider.ESM_Data.CONTENT_URI, null, null, null, "timestamp DESC");
//        boolean isEqual = false;
//
//        // Doing it this way to make sure we dont only check the subject id, which is an ESM also
//        if (survey_data.getCount() > 1) {
//            survey_data.moveToFirst();
//            survey_data.moveToNext();
//            float latest_timestamp = survey_data.getFloat(survey_data.getColumnIndex("timestamp"));
//            if (!survey_data.isClosed()) survey_data.close();
//
//            Calendar cal_today = Calendar.getInstance();
//            cal_today.setTimeInMillis(System.currentTimeMillis());
//            Calendar cal_survey = Calendar.getInstance();
//            cal_survey.setTimeInMillis((long)latest_timestamp);
//
//            isEqual = (cal_today.get(Calendar.YEAR) == cal_survey.get(Calendar.YEAR))
//                    && (cal_today.get(Calendar.MONTH) == cal_survey.get(Calendar.MONTH))
//                    && (cal_today.get(Calendar.DAY_OF_MONTH) == cal_survey.get(Calendar.DAY_OF_MONTH));
//        }
//
//        return isEqual;
//    }
//
//    private void updateCompletedTextVisibility(View rootView) {
//        /*TextView alreadyCompletedText = rootView.findViewById(R.id.text_completed);
//        ImageView alreadyCompletedImg = rootView.findViewById(R.id.img_completed);
//        if (hasFilledSurveyToday()) {
//            alreadyCompletedText.setVisibility(View.VISIBLE);
//            alreadyCompletedImg.setVisibility(View.VISIBLE);
//        } else {
//            alreadyCompletedText.setVisibility(View.GONE);
//            alreadyCompletedImg.setVisibility(View.GONE);
//        }*/
//    }
//
//}

File: app/src/main/java/edu/ucla/cs/er/depressionstudy/SurveyFragment.kt
package edu.ucla.cs.er.depressionstudy

import android.content.ContentValues
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.aware.Aware
import com.aware.Aware_Preferences
import com.aware.ESM
import com.aware.providers.ESM_Provider
import com.google.gson.Gson
import com.quickbirdstudios.surveykit.*
import com.quickbirdstudios.surveykit.result.StepResult
import com.quickbirdstudios.surveykit.result.TaskResult
import com.quickbirdstudios.surveykit.steps.CompletionStep
import com.quickbirdstudios.surveykit.steps.InstructionStep
import com.quickbirdstudios.surveykit.steps.QuestionStep
import com.quickbirdstudios.surveykit.steps.Step
import com.quickbirdstudios.surveykit.survey.SurveyView


class SurveyFragment : Fragment() {
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_survey, container, false)

        var surveyView: SurveyView = rootView!!.findViewById(R.id.survey_view)

        val instructions = InstructionStep(
                title = "Survey",
                text = "You can start the survey by pressing the button below.",
                buttonText = "Start"
        )

        val ageq = QuestionStep(
                title = "Age",
                text = "How old are you today?",
                answerFormat = AnswerFormat.IntegerAnswerFormat()
        )

        val substanceUse = QuestionStep(
                title = "Substance Use",
                text = "In the past 3 months, have you used substances such as crystal meth or injectable drugs not prescribed to you by a physician?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("Yes, methamphetamines"),
                                TextChoice("Yes, injectable drugs not prescribed"),
                                TextChoice("Yes, used both"),
                                TextChoice("Neither"))
                )
        )

        val pwid1 = QuestionStep(
                title = "HIV Risk",
                text = "In the past 3 months, were you in a methadone maintenance program?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("Yes"),
                                TextChoice("No"))
                )
        )

        val pwid2 = QuestionStep(
                title = "HIV Risk",
                text = "In the past 3 months, did you inject cocaine?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("Yes"),
                                TextChoice("No"))
                )
        )
        val pwid3 = QuestionStep(
                title = "HIV Risk",
                text = "In the past 3 months, did you inject heroin?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("Yes"),
                                TextChoice("No"))
                )
        )
        val pwid4 = QuestionStep(
                title = "HIV Risk",
                text = "In the past 3 months, did you use needles?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("Yes"),
                                TextChoice("No"))
                )
        )
        val pwid5 = QuestionStep(
                title = "HIV Risk",
                text = "In the past 3 months, did you share a cooker?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("Yes"),
                                TextChoice("No"))
                )
        )
        val pwid6 = QuestionStep(
                title = "HIV Risk",
                text = "In the past 3 months, did you visit a shooting gallery?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("Yes"),
                                TextChoice("No"))
                )
        )

        val sexBehavior1 = QuestionStep(
                title = "Sexual Behavior and Risk",
                text = "In the last 3 months, how many men have you had sex with?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice(">10"),
                                TextChoice("6-10"),
                                TextChoice("1-5"),
                                TextChoice("0"))
                )
        )

        val sexBehavior2 = QuestionStep(
                title = "Sexual Behavior and Risk",
                text = "In the last 3 months, how many times did you have receptive anal sex (you were the bottom) with a man when he did not use a condom?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("1 or more times"),
                                TextChoice("0 times"))
                )
        )

        val sexBehavior3 = QuestionStep(
                title = "Sexual Behavior and Risk",
                text = "In the last 3 months, how many of your male sex partners were HIV-positive?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("More than 1 HIV+ male partners"),
                                TextChoice("1 HIV+ male partner"),
                                TextChoice("0"),
                                TextChoice("Don't know"))
                )
        )

        val sexBehavior4 = QuestionStep(
                title = "Sexual Behavior and Risk",
                text = "In the last 3 months, how many times did you have insertive anal sex (you were the top) with a man who was HIV-positive when you did not use a condom?",
                answerFormat = AnswerFormat.SingleChoiceAnswerFormat(
                        textChoices = listOf(TextChoice("5 or more times"),
                                TextChoice("0-4 times"))
                )
        )

        val finished = CompletionStep(
                title = "Done!",
                text = "Thank you for filling the survey! Please press the button below to submit the answers and return to the main page.",
                buttonText = "Submit"
        )

        val steps = listOf(instructions, ageq, substanceUse, pwid1, pwid2, pwid3, pwid4, pwid5, pwid6, sexBehavior1, sexBehavior2, sexBehavior3, sexBehavior4, finished)
        val task = NavigableOrderedTask(steps = steps)

        task.setNavigationRule(
                steps[steps.indexOf(substanceUse)].id,
                NavigationRule.ConditionalDirectionStepNavigationRule(
                        resultToStepIdentifierMapper = { input ->
                            when (input) {
                                "Yes, methamphetamines" -> steps[steps.indexOf(sexBehavior1)].id
                                "Neither" -> steps[steps.indexOf(sexBehavior1)].id
                                else -> steps[steps.indexOf(pwid1)].id
                            }
                        }
                )
        )

        task.setNavigationRule(
                steps[steps.indexOf(sexBehavior1)].id,
                NavigationRule.ConditionalDirectionStepNavigationRule(
                        resultToStepIdentifierMapper = { input ->
                            when (input) {
                                "0" -> steps[steps.indexOf(finished)].id
                                else -> steps[steps.indexOf(sexBehavior2)].id
                            }
                        }
                )
        )

        task.setNavigationRule(
                steps[steps.indexOf(sexBehavior3)].id,
                NavigationRule.ConditionalDirectionStepNavigationRule(
                        resultToStepIdentifierMapper = { input ->
                            when (input) {
                                "0" -> steps[steps.indexOf(finished)].id
                                "Don't know" -> steps[steps.indexOf(finished)].id
                                else -> steps[steps.indexOf(sexBehavior4)].id
                            }
                        }
                )
        )

        val configuration = SurveyTheme(
                themeColorDark = ContextCompat.getColor(requireContext(), R.color.colorPrimaryDark),
                themeColor = ContextCompat.getColor(requireContext(), R.color.colorPrimary),
                textColor = ContextCompat.getColor(requireContext(), R.color.black)
        )

        surveyView.onSurveyFinish = { taskResult: TaskResult, reason: FinishReason ->
            if (reason == FinishReason.Completed) {
                var msmScore = 0
                var pwidScore = 0

                val results = taskResult.results.map { res -> res.results.first().stringIdentifier }

                var age = results[1].toInt()
                if (age < 18) {
                    msmScore += 0
                } else if (age < 29) {
                    msmScore += 8
                } else if (age < 41) {
                    msmScore += 5
                } else if (age < 49) {
                    msmScore += 2
                }

                if (results[2].equals("Yes, methamphetamines")
                        || results[2].equals("Yes, used both")) {
                    msmScore += 6
                }

                var nextIndex = 3
                if (results[2].equals("Yes, injectable drugs not prescribed")
                        || results[2].equals("Yes, used both")) {
                    if (age < 30) {
                        pwidScore += 38
                    } else if (age < 40) {
                        pwidScore += 24
                    } else if (age < 50) {
                        pwidScore += 7
                    } else {
                        pwidScore += 2
                    }

                    if (results[3] == "No") {
                        pwidScore += 31
                    }

                    var pwidSubScore = 0
                    if (results[4] == "Yes") {
                        pwidSubScore += 1
                    }
                    if (results[5] == "Yes") {
                        pwidSubScore += 1
                    }
                    if (results[6] == "Yes") {
                        pwidSubScore += 1
                    }
                    if (results[7] == "Yes") {
                        pwidSubScore += 1
                    }
                    if (results[8] == "Yes") {
                        pwidSubScore += 1
                    }

                    if (pwidSubScore == 1) {
                        pwidScore += 7
                    } else if (pwidSubScore == 2) {
                        pwidScore += 21
                    } else if (pwidSubScore == 3) {
                        pwidScore += 24
                    } else if (pwidSubScore == 4) {
                        pwidScore += 24
                    } else if (pwidSubScore == 5) {
                        pwidScore += 31
                    }

                    nextIndex = 9
                }

                if (results[nextIndex] == "0") {
                    nextIndex = -1
                } else if (results[nextIndex] == "1-5") {
                    nextIndex += 1
                } else if (results[nextIndex] == "6-10") {
                    nextIndex += 1
                    msmScore += 4
                } else if (results[nextIndex] == ">10") {
                    nextIndex += 1
                    msmScore += 7
                }

                if (nextIndex > 0) {
                    if (results[nextIndex] == "1 or more times") {
                        msmScore += 10
                    }
                    nextIndex += 1

                    if (results[nextIndex] == "More than 1 HIV+ male partners") {
                        msmScore += 8
                        nextIndex += 1
                    } else if (results[nextIndex] == "1 HIV+ male partner") {
                        msmScore += 4
                        nextIndex += 1
                    } else {
                        nextIndex = -1
                    }

                    if (nextIndex > 0 && results[nextIndex] == "5 or more times") {
                        msmScore += 6
                    }
                }

                val processedResults = resultsToPairs(taskResult.results, steps)
                val storeValue = mapOf("type" to "survey", "msmScore" to msmScore, "pwidScore" to pwidScore, "answers" to processedResults)
                val resultJson = Gson().toJson(storeValue)

                Log.e("logTag", "resultJson: $resultJson")

                val rowData = ContentValues()
                rowData.put(ESM_Provider.ESM_Data.TIMESTAMP, System.currentTimeMillis())
                rowData.put(ESM_Provider.ESM_Data.ANSWER_TIMESTAMP, System.currentTimeMillis())
                rowData.put(ESM_Provider.ESM_Data.DEVICE_ID, Aware.getSetting(context, Aware_Preferences.DEVICE_ID))
                rowData.put(ESM_Provider.ESM_Data.JSON, resultJson)
                rowData.put(ESM_Provider.ESM_Data.STATUS, ESM.STATUS_ANSWERED)

                context?.contentResolver?.insert(ESM_Provider.ESM_Data.CONTENT_URI, rowData)

                val builder: AlertDialog.Builder = AlertDialog.Builder(context!!)
                builder.setMessage("PWID: $pwidScore, MSM: $msmScore")
                        .setTitle("Scores")

                val dialog: AlertDialog = builder.create()
                dialog.show()

                (activity as MainActivity).openTab(0)
            } else {
                surveyView.start(task, configuration)
                if (activity is MainActivity) {
                    (activity as MainActivity).openTab(1)
                }
            }
        }


        surveyView.start(task, configuration)

        return rootView
    }

    fun resultsToPairs(results: List<StepResult>, steps: List<Step>): List<Map<String, String>> {
        fun findStep(result: StepResult): Step? {
            return steps.find { step -> result.results.first().id == step.id }
        }

        fun questionFromStep(step: Step?): String {
            return if (step is QuestionStep) {
                step.text
            } else {
                ""
            }
        }

        fun questionFromResult(result: StepResult): String {
            return questionFromStep(findStep(result))
        }

        val questionAnswerPairs = results.map { result -> mapOf("question" to questionFromResult(result), "answer" to result.results.first().stringIdentifier) }
                .filter { pair -> pair["question"] != "" } // Remove intro / completion steps

        return questionAnswerPairs
    }
}